### PR TITLE
feat(xai): initial xai (grok) implementation

### PR DESCRIPTION
**Pull request by 0xMochan**
Implements xAI (grok) support.

## Notes
This implementation follows very closely to the openai implementation. It seems like xai has modeled their api to match both openai _and_ anthropic (simultaneously). The docs are very unclear on most of the sub-models so the implementation was essentially matched against openai. It might be possible to test alternative BASE_URLs schemas on the openai and anthropic APIs as suggested in the docs via python/js official SDK examples.

The `v1/embeddings` endpoint is documented and lists a model `v1` in the docs. Afaik, this is an internal model not available to the public as it's not listed as available in the xAI console (see photo). Therefore, this is untested (but likely would work if you had access to this model).

image

### CODE REVIEW

File: rig-core/src/providers/xai/embedding.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+use crate::embeddings::{self, EmbeddingError};
+
+use super::{
+    client::xai_api_types::{ApiErrorResponse, ApiResponse},
+    Client,
+};
+
+// ================================================================
+// OpenAI Embedding API
```
Comment (mateobelanger) [inline rig-core/src/providers/xai/embedding.rs:17]:
```suggestion
// xAI Embedding API
```

File: rig-core/src/providers/xai/client.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    embeddings::{self},
+    extractor::ExtractorBuilder,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::{completion::CompletionModel, embedding::EmbeddingModel, EMBEDDING_V1};
+
+// ================================================================
+// Google Gemini Client
```
Comment (mateobelanger) [inline rig-core/src/providers/xai/client.rs:12]:
```suggestion
// xAI Client
```

**Review by mateobelanger — COMMENTED** [review]
Can you add a reference to xAI in the readme's integration section? With the logo ideally!

**Review by cvauclair — CHANGES_REQUESTED** [review]
Nicely done! Only thing left is to clean up the main example (which is pretty messy atm) and addressing @mateobelanger 's comments.

Also, don't forget to add `xAI` to the list of providers in the repo README and the `rig-core/src/lib.rs` lib docstrings!

**Comment by 0xMochan** [comment]
- Added general style comment changes
- Added xai to various READMEs
- Added some printlines to grok example

File: rig-core/README.md [hunk 3]
```diff
@@ -48,5 +50,6 @@ Rig supports the following LLM providers natively:
 - OpenAI
 - Cohere
 - Google Gemini
+- xAI
 Additionally, Rig currently has the following integration sub-libraries:
```
Comment (cvauclair) [inline rig-core/README.md:54]: This is missing a line break (renders as a single line rn)

**Review by cvauclair — CHANGES_REQUESTED** [review]
Just a small thing to fix in the README and it should be good to go!

**Review by cvauclair — APPROVED** [review]

**Pull request merged**